apm-sniffer/bytebuddy-patch/src/test/java/org/apache/skywalking/apm/agent/bytebuddy/SWClassFileLocator.java

```diff
@@ -0,0 +1,174 @@
+    private Thread thread;
+    private int timeoutSeconds = 2;
+    private volatile boolean closed;
+
+    public SWClassFileLocator(Instrumentation instrumentation, ClassLoader classLoader) {
+        this.instrumentation = instrumentation;
+        this.classLoader = classLoader;
+        classLoadingDelegate = ForInstrumentation.ClassLoadingDelegate.ForDelegatingClassLoader.of(classLoader);
+
+        // Use thread instead of ExecutorService here, avoiding conflicts with apm-jdk-threadpool-plugin
+        thread = new Thread(() -> {
+            while (!closed) {
+                try {
+                    ResolutionFutureTask task = queue.poll(5, TimeUnit.SECONDS);
+                    if (task != null) {
+                        try {
+                            Resolution resolution = getResolution(task.getClassName());
+                            task.getFuture().complete(resolution);
+                        } catch (Throwable e) {
+                            task.getFuture().completeExceptionally(e);
+                        }
+                    }
+                } catch (InterruptedException e) {
+                    // ignore interrupted error
+                } catch (Throwable e) {
+                    LOGGER.error(e, "Resolve bytecode of class failure");
+                }
+            }
+        }, "SWClassFileLocator");
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    @Override
+    public Resolution locate(String name) throws IOException {
+        if (!match(name)) {
+            return new Resolution.Illegal(name);
+        }
+        if (closed) {
+            throw new IOException("resolve class failure: closed");
+        }
+        // get class binary representation in a clean thread, avoiding nest calling transformer!
+        ResolutionFutureTask futureTask = new ResolutionFutureTask(name);
+        queue.offer(futureTask);
+        try {
+            return futureTask.getFuture().get(timeoutSeconds, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            throw new IOException("resolve class failure: " + name, e);
+        }
+    }
+
+    private boolean match(String name) {
+        boolean matched = false;
+        for (String typeNameTrait : typeNameTraits) {
+            if (name.contains(typeNameTrait)) {
+                matched = true;
+                break;
+            }
+        }
+        return matched;
+    }
+
+    private Resolution getResolution(String name) throws Exception {
+        SWExtractionClassFileTransformer classFileTransformer = new SWExtractionClassFileTransformer(name);
+        try {
+            instrumentation.addTransformer(classFileTransformer, true);
+            Class aClass = locateClass(name);
+            if (aClass == null) {
+                return new Resolution.Illegal(name);
+            }
+            instrumentation.retransformClasses(new Class[]{aClass});
```

Comment:
What does this `retransformClasses` mean?

Reply:
Trigger re-transforming the target class, and receive bytecode in SWExtractionClassFileTransformer.
This is a way to get bytecode from a class, especially a dynamically generated or enhanced class. Only original bytecode can be read from the classpath, and without dynamically generated classes.